File: src/train/ppo.ts
import * as tf from '@tensorflow/tfjs'
import type { RngFn } from '../sim'
import { randInt } from '../sim'
import type { ActorCriticModel } from './network'
import { evaluateMaskedCategorical } from './actionSampling'
import type { PPOConfig, RolloutBatch } from './types'

export interface PpoUpdateStats {
  policyLoss: number
  valueLoss: number
  entropy: number
  /** "k2"近似KL(0.5 * mean((newLogProb-oldLogProb)^2))。学習が暴走していないかの目安。 */
  approxKl: number
  /** レシオがクリップ範囲外だったサンプルの割合。 */
  clipFraction: number
}

function shuffledIndices(n: number, rng: RngFn): number[] {
  const indices = Array.from({ length: n }, (_, i) => i)
  for (let i = n - 1; i > 0; i--) {
    const j = randInt(rng, i + 1)
    ;[indices[i], indices[j]] = [indices[j], indices[i]]
  }
  return indices
}

/**
 * PPOのクリップ付きサロゲート目的関数による1回の更新(複数epoch・ミニバッチSGD)。
 *
 * `tf.variableGrads(loss)`はvarList省略で呼ぶ — 省略時はプロセス内の全trainable変数が対象になる
 * ため、呼び出し時点でこの`model`以外に学習対象のtrainable変数が存在しないこと(network.tsの
 * ドキュメント参照)。損失計算は`tf.tidy`でラップしてよい(検証済み: `variableGrads`のコールバック内で
 * `tf.tidy`をネストしても、逆伝播に必要な中間tensorはエンジンが正しく保護するため勾配は壊れない) —
 * これにより中間tensor(1ミニバッチあたり十数個)を1つずつ手動disposeせずに済む。
 * ミニバッチのgather結果とgradだけは`tf.tidy`の外で生成される値なので、明示的にdisposeする。
 */
export function runPpoUpdate(
  model: ActorCriticModel,
  optimizer: tf.Optimizer,
  batch: RolloutBatch,
  config: PPOConfig,
  rng: RngFn,
): PpoUpdateStats {
  const steps = batch.steps
  const n = steps.length
  if (n === 0) {
    return { policyLoss: 0, valueLoss: 0, entropy: 0, approxKl: 0, clipFraction: 0 }
  }

  const obsT = tf.tensor2d(steps.map((s) => s.obs))
  const moveMaskT = tf.tensor2d(steps.map((s) => s.moveMask.map((b) => (b ? 1 : 0))))
  const attackMaskT = tf.tensor2d(steps.map((s) => s.attackMask.map((b) => (b ? 1 : 0))))
  const moveActionsT = tf.tensor1d(
    steps.map((s) => s.moveAction),
    'int32',
  )
  const attackActionsT = tf.tensor1d(
    steps.map((s) => s.attackAction),
    'int32',
  )
  const oldLogProbT = tf.tensor1d(steps.map((s) => s.oldLogProb))
  const oldValueT = tf.tensor1d(steps.map((s) => s.value))
  const returnsT = tf.tensor1d(steps.map((s) => s.return))

  // バッチ全体で1回だけ正規化する(epochループの外側)。
  const rawAdvantages = steps.map((s) => s.advantage)
  const advMean = rawAdvantages.reduce((a, b) => a + b, 0) / n
  const advVariance = rawAdvantages.reduce((a, b) => a + (b - advMean) ** 2, 0) / n
  const advStd = Math.sqrt(advVariance)
  const advantagesT = tf.tensor1d(rawAdvantages.map((a) => (a - advMean) / (advStd + 1e-8)))

  const useClippedValueLoss = config.clipValueLoss ?? true
  const accum = { policyLoss: 0, valueLoss: 0, entropy: 0, approxKl: 0, clipFraction: 0, weight: 0 }

  try {
    for (let epoch = 0; epoch < config.epochs; epoch++) {
      const order = shuffledIndices(n, rng)
      for (let start = 0; start < n; start += config.minibatchSize) {
        const idx = order.slice(start, start + config.minibatchSize)
        const idxT = tf.tensor1d(idx, 'int32')

        let batchStats: PpoUpdateStats | undefined

        const { value: lossValue, grads } = tf.variableGrads(() =>
          tf.tidy(() => {
            const obsSlice = tf.gather(obsT, idxT) as tf.Tensor2D
            const moveMaskSlice = tf.gather(moveMaskT, idxT) as tf.Tensor2D
            const attackMaskSlice = tf.gather(attackMaskT, idxT) as tf.Tensor2D
            const moveActionsSlice = tf.gather(moveActionsT, idxT) as tf.Tensor1D
            const attackActionsSlice = tf.gather(attackActionsT, idxT) as tf.Tensor1D
            const oldLogProbSlice = tf.gather(oldLogProbT, idxT) as tf.Tensor1D
            const oldValueSlice = tf.gather(oldValueT, idxT) as tf.Tensor1D
            const returnsSlice = tf.gather(returnsT, idxT) as tf.Tensor1D
            const advantagesSlice = tf.gather(advantagesT, idxT) as tf.Tensor1D

            const { moveLogits, attackLogits, value } = model.forward(obsSlice)
            const moveEval = evaluateMaskedCategorical(moveLogits, moveMaskSlice, moveActionsSlice)
            const attackEval = evaluateMaskedCategorical(attackLogits, attackMaskSlice, attackActionsSlice)
            const newLogProb = tf.add(moveEval.logProbs, attackEval.logProbs) as tf.Tensor1D
            const entropyPerSample = tf.add(moveEval.entropy, attackEval.entropy) as tf.Tensor1D

            const ratio = tf.exp(tf.sub(newLogProb, oldLogProbSlice))
            const surr1 = tf.mul(ratio, advantagesSlice)
            const clippedRatio = tf.clipByValue(ratio, 1 - config.clipRatio, 1 + config.clipRatio)
            const surr2 = tf.mul(clippedRatio, advantagesSlice)
            const policyLoss = tf.neg(tf.mean(tf.minimum(surr1, surr2))) as tf.Scalar

            const valuePred = value.reshape([value.shape[0]]) as tf.Tensor1D
            let valueLossTensor: tf.Scalar
            if (useClippedValueLoss) {
              const delta = tf.clipByValue(tf.sub(valuePred, oldValueSlice), -config.clipRatio, config.clipRatio)
              const valueClipped = tf.add(oldValueSlice, delta)
              const lossUnclipped = tf.square(tf.sub(valuePred, returnsSlice))
              const lossClipped = tf.square(tf.sub(valueClipped, returnsSlice))
              valueLossTensor = tf.mean(tf.maximum(lossUnclipped, lossClipped)) as tf.Scalar
            } else {
              valueLossTensor = tf.mean(tf.square(tf.sub(valuePred, returnsSlice))) as tf.Scalar
            }

            const entropyMean = tf.mean(entropyPerSample) as tf.Scalar
            const total = tf.add(
              tf.add(policyLoss, tf.mul(valueLossTensor, config.valueLossCoef)),
              tf.neg(tf.mul(entropyMean, config.entropyCoef)),
            ) as tf.Scalar

            const approxKl = tf.mul(tf.mean(tf.square(tf.sub(newLogProb, oldLogProbSlice))), 0.5) as tf.Scalar
            const clipFrac = tf.mean(
              tf.cast(tf.greater(tf.abs(tf.sub(ratio, 1)), config.clipRatio), 'float32'),
            ) as tf.Scalar

            batchStats = {
              policyLoss: policyLoss.dataSync()[0],
              valueLoss: valueLossTensor.dataSync()[0],
              entropy: entropyMean.dataSync()[0],
              approxKl: approxKl.dataSync()[0],
              clipFraction: clipFrac.dataSync()[0],
            }

            return total
          }),
        )

        optimizer.applyGradients(grads)

        const weight = idx.length
        if (batchStats) {
          accum.policyLoss += batchStats.policyLoss * weight
          accum.valueLoss += batchStats.valueLoss * weight
          accum.entropy += batchStats.entropy * weight
          accum.approxKl += batchStats.approxKl * weight
          accum.clipFraction += batchStats.clipFraction * weight
          accum.weight += weight
        }

        lossValue.dispose()
        for (const key in grads) grads[key].dispose()
        idxT.dispose()
      }
    }
  } finally {
    tf.dispose([obsT, moveMaskT, attackMaskT, moveActionsT, attackActionsT, oldLogProbT, oldValueT, returnsT, advantagesT])
  }

  const w = Math.max(1, accum.weight)
  return {
    policyLoss: accum.policyLoss / w,
    valueLoss: accum.valueLoss / w,
    entropy: accum.entropy / w,
    approxKl: accum.approxKl / w,
    clipFraction: accum.clipFraction / w,
  }
}

File: training-py/mayhem_rl/ppo.py
"""PPO clipped-surrogate update. Mirrors src/train/ppo.ts. Preserve specific details deliberately
during any future edits: advantages are normalized ONCE per batch, OUTSIDE the epoch loop (easy to
accidentally "fix" into per-epoch normalization) and the value loss defaults to the PPO2-style
clipped MSE (see `PPOConfig.clip_value_loss`).

Unlike the TF.js side, PyTorch's autograd manages tensor memory automatically — there is no
manual-disposal discipline needed here (that was the real correctness risk on the TS side; it
does not carry over).
"""

from __future__ import annotations

import random
from dataclasses import dataclass

import torch
import torch.nn.functional as F

from .action_sampling import evaluate_masked_categorical
from .network import ActorCriticNetwork


@dataclass
class PPOConfig:
    clip_ratio: float = 0.2
    epochs: int = 4
    minibatch_size: int = 256
    value_loss_coef: float = 0.5
    entropy_coef: float = 0.01
    clip_value_loss: bool = True
    learning_rate: float = 3e-4
    max_grad_norm: float | None = None


@dataclass
class RolloutStep:
    obs: list[float]
    move_mask: list[bool]
    attack_mask: list[bool]
    ability_mask: list[bool]
    move_action: int
    attack_action: int
    ability_action: int
    old_log_prob: float
    value: float
    advantage: float
    return_: float


@dataclass
class PpoUpdateStats:
    policy_loss: float
    value_loss: float
    entropy: float
    # "k2" approximate KL (0.5 * mean((newLogProb-oldLogProb)^2)) -- a rough gauge of how far the
    # update moved the policy.
    approx_kl: float
    clip_fraction: float


def run_ppo_update(
    network: ActorCriticNetwork,
    optimizer: torch.optim.Optimizer,
    steps: list[RolloutStep],
    config: PPOConfig,
    device: torch.device,
    rng: random.Random,
) -> PpoUpdateStats:
    n = len(steps)
    if n == 0:
        return PpoUpdateStats(0.0, 0.0, 0.0, 0.0, 0.0)

    obs_t = torch.tensor([s.obs for s in steps], dtype=torch.float32, device=device)
    move_mask_t = torch.tensor([[1.0 if b else 0.0 for b in s.move_mask] for s in steps], dtype=torch.float32, device=device)
    attack_mask_t = torch.tensor(
        [[1.0 if b else 0.0 for b in s.attack_mask] for s in steps], dtype=torch.float32, device=device
    )
    ability_mask_t = torch.tensor(
        [[1.0 if b else 0.0 for b in s.ability_mask] for s in steps], dtype=torch.float32, device=device
    )
    move_actions_t = torch.tensor([s.move_action for s in steps], dtype=torch.long, device=device)
    attack_actions_t = torch.tensor([s.attack_action for s in steps], dtype=torch.long, device=device)
    ability_actions_t = torch.tensor([s.ability_action for s in steps], dtype=torch.long, device=device)
    old_log_prob_t = torch.tensor([s.old_log_prob for s in steps], dtype=torch.float32, device=device)
    old_value_t = torch.tensor([s.value for s in steps], dtype=torch.float32, device=device)
    returns_t = torch.tensor([s.return_ for s in steps], dtype=torch.float32, device=device)

    raw_advantages = [s.advantage for s in steps]
    adv_mean = sum(raw_advantages) / n
    adv_var = sum((a - adv_mean) ** 2 for a in raw_advantages) / n
    adv_std = adv_var**0.5
    advantages_t = torch.tensor(
        [(a - adv_mean) / (adv_std + 1e-8) for a in raw_advantages], dtype=torch.float32, device=device
    )

    accum = {"policy_loss": 0.0, "value_loss": 0.0, "entropy": 0.0, "approx_kl": 0.0, "clip_fraction": 0.0, "weight": 0}

    for _epoch in range(config.epochs):
        order = list(range(n))
        rng.shuffle(order)
        for start in range(0, n, config.minibatch_size):
            idx = order[start : start + config.minibatch_size]
            idx_t = torch.tensor(idx, dtype=torch.long, device=device)

            obs_slice = obs_t[idx_t]
            move_mask_slice = move_mask_t[idx_t]
            attack_mask_slice = attack_mask_t[idx_t]
            ability_mask_slice = ability_mask_t[idx_t]
            move_actions_slice = move_actions_t[idx_t]
            attack_actions_slice = attack_actions_t[idx_t]
            ability_actions_slice = ability_actions_t[idx_t]
            old_log_prob_slice = old_log_prob_t[idx_t]
            old_value_slice = old_value_t[idx_t]
            returns_slice = returns_t[idx_t]
            advantages_slice = advantages_t[idx_t]

            move_logits, attack_logits, ability_logits, value = network(obs_slice)
            move_log_prob, move_entropy = evaluate_masked_categorical(move_logits, move_mask_slice, move_actions_slice)
            attack_log_prob, attack_entropy = evaluate_masked_categorical(
                attack_logits, attack_mask_slice, attack_actions_slice
            )
            ability_log_prob, ability_entropy = evaluate_masked_categorical(
                ability_logits, ability_mask_slice, ability_actions_slice
            )
            new_log_prob = move_log_prob + attack_log_prob + ability_log_prob
            entropy_per_sample = move_entropy + attack_entropy + ability_entropy

            ratio = (new_log_prob - old_log_prob_slice).exp()
            surr1 = ratio * advantages_slice
            clipped_ratio = torch.clamp(ratio, 1 - config.clip_ratio, 1 + config.clip_ratio)
            surr2 = clipped_ratio * advantages_slice
            policy_loss = -torch.min(surr1, surr2).mean()

            value_pred = value.squeeze(-1)
            if config.clip_value_loss:
                delta = torch.clamp(value_pred - old_value_slice, -config.clip_ratio, config.clip_ratio)
                value_clipped = old_value_slice + delta
                loss_unclipped = (value_pred - returns_slice) ** 2
                loss_clipped = (value_clipped - returns_slice) ** 2
                value_loss = torch.max(loss_unclipped, loss_clipped).mean()
            else:
                value_loss = F.mse_loss(value_pred, returns_slice)

            entropy_mean = entropy_per_sample.mean()
            total_loss = policy_loss + value_loss * config.value_loss_coef - entropy_mean * config.entropy_coef

            optimizer.zero_grad()
            total_loss.backward()
            if config.max_grad_norm is not None:
                torch.nn.utils.clip_grad_norm_(network.parameters(), config.max_grad_norm)
            optimizer.step()

            with torch.no_grad():
                approx_kl = 0.5 * ((new_log_prob - old_log_prob_slice) ** 2).mean()
                clip_fraction = ((ratio - 1.0).abs() > config.clip_ratio).float().mean()

            weight = len(idx)
            accum["policy_loss"] += policy_loss.item() * weight
            accum["value_loss"] += value_loss.item() * weight
            accum["entropy"] += entropy_mean.item() * weight
            accum["approx_kl"] += approx_kl.item() * weight
            accum["clip_fraction"] += clip_fraction.item() * weight
            accum["weight"] += weight

    w = max(1, accum["weight"])
    return PpoUpdateStats(
        policy_loss=accum["policy_loss"] / w,
        value_loss=accum["value_loss"] / w,
        entropy=accum["entropy"] / w,
        approx_kl=accum["approx_kl"] / w,
        clip_fraction=accum["clip_fraction"] / w,
    )
